Classification/FeatureSelection/src/main/java/org/tribuo/classification/fs/FS_Wrapper_Approaches/Optimizers/CuckooSearchOptimizer.java

```diff
@@ -0,0 +1,157 @@
+ *
+ * L. A. M. Pereira et al.
+ * "A Binary Cuckoo Search and its Application for Feature Selection", 2014.
+ * </pre>
+ */
+public class CuckooSearchOptimizer implements FeatureSelector<Label> {
+    private final TransferFunction transferFunction;
+    private final double stepSizeScaling;
+    private final double lambda;
+    private final double worstNestProbability;
+    private final double delta;
+    private final int populationSize;
+    private int [][] setOfSolutions;
+    private final int maxIteration;
+
+    /**
+     * The default constructor for feature selection based on Cuckoo Search Algorithm
+     */
+    public CuckooSearchOptimizer() {
+        this.transferFunction = TransferFunction.TFunction_V2;
+        this.populationSize = 50;
+        this.stepSizeScaling = 2d;
+        this.lambda = 2d;
+        this.worstNestProbability = 0.1d;
+        this.delta = 1.5d;
+        this.maxIteration = 30;
+    }
+    
+    /**
+     * Constructs the wrapper feature selection based on cuckoo search algorithm
+     * @param transferFunction The transfer function to convert continuous values to binary ones
+     * @param populationSize The size of the solution in the initial population
+     * @param maxIteration The number of times that is used to enhance generation
+     */
+    public CuckooSearchOptimizer(TransferFunction transferFunction, int populationSize, int maxIteration) {
+        this.transferFunction = transferFunction;
+        this.populationSize = populationSize;
+        this.stepSizeScaling = 2d;
+        this.lambda = 2d;
+        this.worstNestProbability = 1.5d;
+        this.delta = 1.5d;
+        this.maxIteration = maxIteration;
+    }
+
+    /**
+     * @param transferFunction The transfer function to convert continuous values to binary ones
+     * @param populationSize The size of the solution in the initial population
+     * @param stepSizeScaling The cuckoo step size
+     * @param lambda The lambda of the levy flight function
+     * @param worstNestProbability The fraction of the nests to be abandoned
+     * @param delta The delta that is used in the abandon nest function
+     * @param maxIteration The number of times that is used to enhance generation
+     */
+    public CuckooSearchOptimizer(TransferFunction transferFunction, int populationSize, double stepSizeScaling, double lambda, double worstNestProbability, double delta, int maxIteration) {
+        this.transferFunction = transferFunction;
+        this.populationSize = populationSize;
+        this.stepSizeScaling = stepSizeScaling;
+        this.lambda = lambda;
+        this.worstNestProbability = worstNestProbability;
+        this.delta = delta;
+        this.maxIteration = maxIteration;
+    }
+
+    /**
+     * @param totalNumberOfFeatures The number of features in the given dataset
+     * @return The population of subsets of selected features
+     */
+    private int[][] GeneratePopulation(int totalNumberOfFeatures) {

```

Comment:
As mentioned elsewhere this method should accept a `SplittableRandom` rather than make a fresh RNG each time.